document/json/encoder_test.go

```diff
@@ -0,0 +1,82 @@
+package json_test
+
+import (
+	"github.com/aws/smithy-go/document/internal/serde"
+	"github.com/aws/smithy-go/document/json"
+	"github.com/google/go-cmp/cmp"
+	"testing"
+	"time"
+)
+
+func TestEncoder_Encode(t *testing.T) {
+	t.Run("Object", func(t *testing.T) {
+		for name, tt := range sharedObjectTests {
+			t.Run(name, func(t *testing.T) {
+				testEncode(t, tt)
+			})
+		}
+	})
+	t.Run("Array", func(t *testing.T) {
+		for name, tt := range sharedArrayTestCases {
+			t.Run(name, func(t *testing.T) {
+				testEncode(t, tt)
+			})
+		}
+	})
+	t.Run("Number", func(t *testing.T) {
+		for name, tt := range sharedNumberTestCases {
+			t.Run(name, func(t *testing.T) {
+				testEncode(t, tt)
+			})
+		}
+	})
+	t.Run("String", func(t *testing.T) {
+		for name, tt := range sharedStringTests {
+			t.Run(name, func(t *testing.T) {
+				testEncode(t, tt)
+			})
+		}
+	})
+}
+
+func TestNewEncoderUnsupportedTypes(t *testing.T) {
+	type customTime time.Time
+
+	cases := []interface{}{
+		time.Now().UTC(),
+		customTime(time.Now().UTC()),
+	}
```

Comment:
Should this include cases for functions and channels?

Reply:
The current behavior for channels and functions is to completely skip them and not serialize them (with no error returned), which is the behavior of the `encoding/json` library. I feel like it makes sense to at least attempt to match that behavior for those types. Time is a bit of a special case since we want to avoid making any sort of opinionated stance on their handling at this time, whereas channels and functions would more then likely never be supported.